hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/statemachine/DatanodeStateMachine.java

```diff
@@ -93,7 +95,9 @@
    *                     enabled
    */
   public DatanodeStateMachine(DatanodeDetails datanodeDetails,
-      Configuration conf, CertificateClient certClient) throws IOException {
+      Configuration conf, CertificateClient certClient,
```

Comment:
It also feels unfortunate that we have to pass this back-reference. StateMachine should not really know about HddsDataNodeService. Now we have a circular dependency.

Can we instead pass a callback via method reference that should be invoked when termination is necessary?

Reply:
Done..